pageserver/src/consumption_metrics/upload.rs

```diff
@@ -0,0 +1,264 @@
+use consumption_metrics::{EventChunk, IdempotencyKey, CHUNK_SIZE};
+use tokio_util::sync::CancellationToken;
+use tracing::Instrument;
+
+use super::{Cache, RawMetric, RawMetricExt};
+
+#[tracing::instrument(skip_all, fields(metrics_total = %metrics.len()))]
+pub(super) async fn upload_metrics(
+    client: &reqwest::Client,
+    metric_collection_endpoint: &reqwest::Url,
+    cancel: &CancellationToken,
+    node_id: &str,
+    metrics: &[RawMetric],
+    cached_metrics: &mut Cache,
+) -> anyhow::Result<()> {
+    let mut uploaded = 0;
+    let mut failed = 0;
+
+    let started_at = std::time::Instant::now();
+
+    for res in serialize_in_chunks(CHUNK_SIZE, metrics, node_id) {
+        let (chunk, body) = res?;
+
+        let event_bytes = body.len();
+
+        let res = upload(client, metric_collection_endpoint, body, cancel)
+            .instrument(tracing::info_span!(
+                "upload",
+                %event_bytes,
+                uploaded,
+                total = metrics.len(),
+            ))
+            .await;
+
+        match res {
+            Ok(()) => {
+                for (curr_key, curr_val) in chunk {
+                    cached_metrics.insert(*curr_key, *curr_val);
+                }
+                uploaded += chunk.len();
+            }
+            Err(_) => {
+                // failure(s) have already been logged
+                //
+                // however this is an inconsistency: if we crash here, we will start with the
+                // values as uploaded. in practice, the rejections no longer happen.
+                failed += chunk.len();
+            }
+        }
+    }
+
+    let elapsed = started_at.elapsed();
+
+    tracing::info!(
+        uploaded,
+        failed,
+        elapsed_ms = elapsed.as_millis(),
+        "done sending metrics"
+    );
+
+    Ok(())
+}
+
+// The return type is quite ugly, but we gain testability in isolation
+fn serialize_in_chunks<'a, F>(
+    chunk_size: usize,
+    input: &'a [RawMetric],
+    factory: F,
+) -> impl Iterator<Item = Result<(&'a [RawMetric], bytes::Bytes), serde_json::Error>> + 'a
+where
+    F: KeyGen<'a> + 'a,
+{
+    use bytes::BufMut;
+
+    // write to a BytesMut so that we can cheaply clone the frozen Bytes for retries
+    let mut buffer = bytes::BytesMut::new();
+    let mut chunks = input.chunks(chunk_size);
+    let mut events = Vec::new();
```

Comment:
nit: `events` could do with a comment explaining that it is a scratch space re-used each time around the loop to minimize allocations of RawMetric's idempotency key.  It took a couple of reads of this function for me understand what was going on here.

Reply:
Yeah, this has been sort of only in discussions with me and @conradludgate, he showed this first and I am not really able to replicate the same with the `Name` not being able to implement `Default`, but it's close enough.

At least this monstrosity of allocation reuse has tests!

Reply:
19867e45a